### DISCUSSION TITLE
How do you specify  an alternate?

### DISCUSSION THREAD

**Discussion by dodyg** [OrchardCMS/OrchardCore#15964] [Q&A]
I need to specify a menu in two different ways in my template, one for large screen and one for mobile. The structure of the menu is the same for both scenarios. They just have different css classes.

I have this normal menu
```
   {% shape "menu", alias: "alias:main-menu", cache_id: "main-menu", cache_fixed_duration: "00:05:00", cache_tag: "alias:main-menu" %}
```
and I am trying to specify the mobile menu but **i cannot make it work**
```
   {% shape "menu", alternate: "Mobile", alias: "alias:main-menu" %}

```
I want the shape to use the template "Menu-Mobile" template but I can't make it work.

I tried the follow values for `alternate` attribute:

* Mobile
* Menu_Mobile
* Menu-Mobile

![alternate](https://user-images.githubusercontent.com/1761249/70889266-dc79ce80-1fea-11ea-831d-da938918657d.png)

**Comment by dodyg** [comment]
OK found it. I have to use the value `Menu__Mobile` as the value for `alternate` to get it to use `Menu-Mobile.cshtml`.

So it is now

```
{% shape "menu", alternate: "Menu__Mobile", alias: "alias:main-menu" %}
```

**Comment by dodyg** [comment]
Now I am trying to figure out how to get `Menu-Mobile.cshtml` to use `MenuItem-Mobile.cshtml`. Right now it keeps using `MenuItem.cshtml`.

This is `Menu-Mobile.cshtml`

```
@{
    TagBuilder tag = Tag(Model, "ul");
    tag.AddCssClass("uk-nav");

    foreach (var item in Model.Items)
    {
        tag.InnerHtml.AppendHtml(await DisplayAsync(item));
    }
}

@tag
```

It needs to use `MenuItem-Mobile.cshtml` next but it's not clear how to do that from C# code.

From this [example](https://github.com/OrchardCMS/OrchardCore/tree/dev/src/OrchardCore.Themes/TheAdmin/Views), there doesn't seem to be  a need to anything special but yet it doesn't work.

**Comment by dodyg** [comment]
Modify the model meta data doesn't work

`Menu-Mobile.cshml`

```
@{
    Model.Metadata.Alternates.Clear();
    Model.Metadata.Alternates.Add("MenuItem__Mobile");

    TagBuilder tag = Tag(Model, "ul");
    tag.AddCssClass("uk-nav");

    foreach (var item in Model.Items)
    {
        tag.InnerHtml.AppendHtml(await DisplayAsync(item));
    }

}

@tag

```

**Comment by deanmarcussen** [comment]
What’s the crash?
Code looks good to me.
Probably don’t need to clear the alternates, just add the new one and it should pick it up

**Comment by dodyg** [comment]
> What’s the crash?

It was my dumb mistake. I wrote ` Model.MetaData` instead of `Model.Metadata`. It still doesn't pick `MenuItem-Mobile.cshtml` though.

I tried `Menu-Mobile.cshtml`

```
@{
    Model.Metadata.Alternates.Add("MenuItem__Mobile");

    TagBuilder tag = Tag(Model, "ul");
    tag.AddCssClass("uk-nav");

    foreach (var item in Model.Items)
    {
        tag.InnerHtml.AppendHtml(await DisplayAsync(item));
    }

}

@tag
```

And it still doesn't pick up `MenuItem-Mobile.cshtml`. It picks up `MenuItem.cshtml`.

**Comment by dodyg** [comment]
OK the correct way for is to do this `Menu-Mobile.cshtml`

```
@{
    TagBuilder tag = Tag(Model, "ul");
    tag.AddCssClass("uk-nav");

    foreach (var item in Model.Items)
    {
        item.Metadata.Alternates.Clear();
        item.Metadata.Alternates.Add("MenuItem__Mobile");
        tag.InnerHtml.AppendHtml(await DisplayAsync(item));
    }
}

@tag

```
This one picks up `MenuItem-Mobile.cshtml`.

**Comment by dodyg** [comment]
The next thing is to figure out how `MenuItem-Mobile.cshtml` to pick up `MenuItemLink-Mobile-LinkMenuItem.cshtml`.

This is `MenuItem-Mobile.cshtml`.

```
@{
    int level = (int)Model.Level;
    TagBuilder tag = Tag(Model, "li");

    if ((bool)Model.HasItems)
    {
        tag.AddCssClass("uk-parent");
    }

    // Morphing the shape to keep Model untouched
    Model.Metadata.Alternates.Clear();
    Model.Metadata.Type = "MenuItemLink";

    tag.InnerHtml.AppendHtml(await DisplayAsync(Model));

    if ((bool)(Model.HasItems))
    {
        TagBuilder parentTag = Tag(Model, "ul");
        parentTag.AddCssClass("uk-nav-sub");

        foreach (var item in Model.Items)
        {
            item.Level = level + 1;
            item.ParentTag = parentTag;
            parentTag.InnerHtml.AppendHtml(await DisplayAsync(item));
        }

        tag.InnerHtml.AppendHtml(parentTag);
    }
}

@tag
```

**Comment by dodyg** [comment]
Modifying `MenuItem-Mobile.cshtml` works partially

```
@{
    int level = (int)Model.Level;
    TagBuilder tag = Tag(Model, "li");

    if ((bool)Model.HasItems)
    {
        tag.AddCssClass("uk-parent-2");
    }

    // Morphing the shape to keep Model untouched
    Model.Metadata.Alternates.Clear();
    Model.Metadata.Type = "MenuItemLink";

    tag.InnerHtml.AppendHtml(await DisplayAsync(Model));

    if ((bool)(Model.HasItems))
    {
        TagBuilder parentTag = Tag(Model, "ul");
        parentTag.AddCssClass("uk-nav-sub");

        foreach (var item in Model.Items)
        {
            item.Metadata.Alternates.Clear();
            item.Metadata.Alternates.Add("MenuItemLink__Mobile__LinkMenuItem");
            item.Level = level + 1;
            item.ParentTag = parentTag;
            parentTag.InnerHtml.AppendHtml(await DisplayAsync(item));
        }

        tag.InnerHtml.AppendHtml(parentTag);
    }
}

@tag

MOBILE
```

The problem is sometimes it picks up `MenuItemLink-Mobile-LinkMenuItem.cshtml` and sometimes it picks up `MenuItemLink-LinkMenuItem.cshtml`. 

When level is 0, it uses `MenuItemLink-LinkMenuItem.cshtml` and when level is 1, it uses `MenuItemLink-Mobile-LinkMenuItem.cshtml`.

I suspect this code ` tag.InnerHtml.AppendHtml(await DisplayAsync(Model));` is calling into `MenuItemLink-LinkMenuItem.cshtml`. So far it resists various types of modications to `Model.Metadata`.

So right now the produced list is below. I modified `MenuItemLink-LinkMenuItem.cshtml` to add `desktop="true"` attribute to the rendering and you see it shows up below.

```
<ul class="uk-nav menu">
    <li class="uk-parent">
        <a desktop="true">About Us<b class="caret"></b></a>
        <ul class="uk-nav-sub">
            <li>
                <a href="/about-us">Who we are</a>
            </li>
            <li>
                <a href="/about-us/ciss-egypt">CISS Egypt</a>
            </li>
            <li>
                <a href="/about-us/the-project">The Project</a>
            </li>
        </ul>
    </li>
    <li class="uk-parent">
        <a desktop="true">Discover New Valley<b class="caret"></b></a>
        <ul class="uk-nav-sub">
            <li>
                <a href="/new-valley">New Valley in a nutshell</a>
            </li>
            <li>
                <a href="/new-valley/how-to-reach">How to reach</a>
            </li>
            <li>
                <a href="/new-valley/businesses">Businesses</a>
            </li>
            <li>
                <a href="/new-valley/tourism">Tourism</a>
            </li>
        </ul>
    </li>
    <li>
        <a href="/catalog">Product Catalog</a>
    </li>
    <li>
        <a href="/publications/">Publications</a>
    </li>
    <li class="uk-parent">
        <a desktop="true">MSEI<b class="caret"></b></a>
        <ul class="uk-nav-sub">
            <li>
                <a href="/msei">About MSEI</a>
            </li>
        </ul>
    </li>
</ul>
```

**Comment by ns8482e** [comment]
> From this [example](https://github.com/OrchardCMS/OrchardCore/tree/dev/src/OrchardCore.Themes/TheAdmin/Views), there doesn't seem to be a need to anything special but yet it doesn't work.

The "catch" is the menu name. `Navigation-admin.cshtml` works without need to do anything special because, alternate for menu name ( i.e. `admin`) is defined out of the box.

In your case the alternate would be `Menu-MainMenu.cshtml` and `Menu-Mobile.cshtml` if you define two separate menus `MainMenu` and `Mobile`
